core/src/nfc_config.c

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (C) 2022 Jolla Ltd.
+ * Copyright (C) 2022 Slava Monich <slava.monich@jolla.com>
+ *
+ * You may use this file under the terms of BSD license as follows:
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *   1. Redistributions of source code must retain the above copyright
+ *      notice, this list of conditions and the following disclaimer.
+ *   2. Redistributions in binary form must reproduce the above copyright
+ *      notice, this list of conditions and the following disclaimer in the
+ *      documentation and/or other materials provided with the distribution.
+ *   3. Neither the names of the copyright holders nor the names of its
+ *      contributors may be used to endorse or promote products derived
+ *      from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "nfc_config.h"
+
+G_DEFINE_INTERFACE(NfcConfigurable, nfc_configurable, G_TYPE_OBJECT)
+
+/*==========================================================================*
+ * Interface
+ *==========================================================================*/
+
+const char* const*
+nfc_config_get_keys(
+    NfcConfigurable* self)
+{
+    return G_LIKELY(self) ?
+        NFC_CONFIGURABLE_GET_IFACE(self)->get_keys(self) :
+        NULL;
+}
+
+GVariant*
+nfc_config_get_value(
+    NfcConfigurable* self,
+    const char* key)
+{
+    if (G_LIKELY(self) && G_LIKELY(key)) {
+        GVariant* val = NFC_CONFIGURABLE_GET_IFACE(self)->get_value(self, key);
+
+        if (val) {
+            /* Make sure we return a full reference */
+            return g_variant_take_ref(val);
+        }
+    }
+    return NULL;
+}
+
+gboolean
+nfc_config_set_value(
+    NfcConfigurable* self,
+    const char* key,
+    GVariant* value)
+{
+    /* Consume weird floating references */
+    GVariant* ref = value ? g_variant_ref_sink(value) : NULL;
```

Comment:
While it is pretty much guaranteed that ref ==value, doing this and still continuing to use "value" looks odd.

Also, setting NULL value seems to be ok, but at least I did not spot what it means being documented. Resets back to defaults, or?

Reply:
When I was writing this, I was thinking like this: if value is non-NULL, we take a ref (or converting the floating one into a normal one). Then in the end, if we've taken the ref, release it. But instead of writing `if (ref) g_variant_unref(ref)` I wrong `if (ref) g_variant_unref(value)` 🙂 I'll make it more consistent.

The NULL behavior is described in `core/include/nfc_config.h`

```
 * NULL value passed to NfcConfigurable::set_value() must be interpreted
 * as the default value.
```

Reply:
Reworked it to drop the unnecessary `ref` variable.